### ISSUE TITLE
Batch the build-and-send instructions into chunks in `solana-tokens distribute-spl-tokens`

### ISSUE THREAD

**Issue by brooksprumo** [solana-labs/solana#27318]
#### Problem

Initiating a feature proposal requires distributing spl tokens to all the validators. On testnet, as of this writing, there are ~3100 nodes. The code currently builds all the transactions, then attempts to send them all. These transactions all fail due to expired blockhashes.

There are multiple RPC calls in this path, and some in loops. So for any medium sized number of accounts, these rpc calls will be rate-limited and timeout the blockhashes. The biggest offender is getting fees, which is used to check if the user has enough SOL in their wallet to distribute the tokens.

#### Proposed Solution

Instead of building all the transactions, then looping over all of them for fees, then trying to send them all; do all this in chunks so that when it comes time to send the transaction, the blockhash isn't old.

May also require (1) creating idempotent accounts, and (2) only querying for a single fee and multiplying by the number of accounts. This will not necessarily be the correct fee, but it will be an upper bound.

**Comment by eloylp** [comment]
Hello :wave: newcomer here. I would like to work on this one. If that is possible, feel free to assign myself to it :+1:

**Comment by eloylp** [comment]
Hello ! After studying a bit the code, i have the following thoughts/doubts i would like to share with you, so we can decide how to move forward. Please, i would be probably doing some assumptions here, correct me at any point.

#### Batch the processing of allocations

Initially, i thought it would be a good idea to batch the entire [distribute_allocations](https://github.com/solana-labs/solana/blob/master/tokens/src/commands.rs#L609) function, which is currently called once:

```rust
pub fn process_allocations(
    client: &RpcClient,
    args: &DistributeTokensArgs,
    exit: Arc<AtomicBool>,
) -> Result<Option<usize>, Error> {
// ...    
distribute_allocations(client, &mut db, &allocations, args, exit.clone())?;
// ...
}
```

Batching the above function would result in something similar to:

```rust
pub fn process_allocations(
    client: &RpcClient,
    args: &DistributeTokensArgs,
    exit: Arc<AtomicBool>,
) -> Result<Option<usize>, Error> {
// ...
allocations
    .chunks(args.batch_size.unwrap_or(allocations.len()))
    .try_for_each(|alloc_chunk| {
        distribute_allocations(client, &mut db, alloc_chunk, args, exit.clone())                       
    })?;
// ...
}
```

In the above code, `args.batch_size` would be added to `clap` args only for the [distribute-spl-tokens subcommand](https://github.com/solana-labs/solana/blob/9a57c64f21ed94e3afbb6c495d5f65c8a799f420/tokens/src/arg_parser.rs#L278), with a wise default (which one would be a good one?) . So the user could set the  batch size via the CLI if necessary. If the argument its not present, it would default to `allocations.len()` , which means no batching. In case something goes wrong in any iteration, the program will return immediately. So preserving all the original behavior here.

With this implementation, we would be limiting the amount of `allocations` that the underlying [build_messages](https://github.com/solana-labs/solana/blob/master/tokens/src/commands.rs#L442) and [send_messages](https://github.com/solana-labs/solana/blob/master/tokens/src/commands.rs#L459) functions needs to process together at once, probably avoiding the outdated blockhash problem, as messages would be sent before that happens.

But not so hasty ! this solution its not as great as i thought. Between the [build_messages](https://github.com/solana-labs/solana/blob/master/tokens/src/commands.rs#L442) and [send_messages](https://github.com/solana-labs/solana/blob/master/tokens/src/commands.rs#L459) function calls, we have the [balance check](https://github.com/solana-labs/solana/blob/master/tokens/src/commands.rs#L453-L457) /fees calculation. This is the responsible of checking if the fee payer and token account have enough balance in order to satisfy the entire operation. That means if we batch the operation at the proposed abstraction level, we would be only doing the balance check per each chunk of `allocations`. What i think would be the problem with this ? well, i think it could leave an operation half done if there is no sufficient balance to process all allocations from all batches. But, its also true that in such case, the user have tools to get the diff, [as from the `README.md`](https://github.com/solana-labs/solana/blob/b8837c04ec3976c9c16d028fbee86f87823fb97f/tokens/README.md#calculate-what-tokens-should-be-sent-1):

> ### Calculate what tokens should be sent
>
> As with SOL, you can List the differences between a list of expected distributions and the record of what transactions have already been sent using the `--dry-run` parameter, or `solana-tokens balances`

So, here are what i consider the possible ways to move forward:

1. Go with the initial idea. Maybe add a bit of information about this behavior in the readme, suggesting users to do `--dry-run` operations first, so if they are out of funds, they will now before sending any message. In any case, if the process fails for whatever reason, they should do a `---dry-run` before applying again the same input CSV (correct me if am wrong).
2. Try to move the balance checks before processing all the `allocation` batches, so only applying the proposed batching code to the [build_messages](https://github.com/solana-labs/solana/blob/master/tokens/src/commands.rs#L442) and [send_messages](https://github.com/solana-labs/solana/blob/master/tokens/src/commands.rs#L459) functions. That would preserve the current behavior (notifying it they are out of funds before sending any message). This needs to be studied with care though, as for checking the [SPL token balances](https://github.com/solana-labs/solana/blob/master/tokens/src/commands.rs#L454), [we depend on the previously created accounts](https://github.com/solana-labs/solana/blob/master/tokens/src/commands.rs#L450-L454), by the [build_messages](https://github.com/solana-labs/solana/blob/master/tokens/src/commands.rs#L442) function, which makes RPC calls, which would be probably needed to do twice then (balance check + build_messages).
3. Do (1) , solve the issues that could be currently causing and then, try (2).
4. Other solutions that i am not taking into account. Happy to hear from you :)

I was taking a look on the [the current testing conventions](https://github.com/solana-labs/solana/blob/master/tokens/src/commands.rs#L1260-L1266). Do you think it would be approachable to setup an automated test for the spl token distribution feature (exercising the batching code) in a similar way ?

#### Relaxing the pressure on the RPC server

As suggested, looks like we could change the current [get_fees_for_messages](https://github.com/solana-labs/solana/blob/9a57c64f21ed94e3afbb6c495d5f65c8a799f420/tokens/src/commands.rs#L736) function:

```rust
pub fn get_fees_for_messages(messages: &[Message], client: &RpcClient) -> Result<u64, Error> { 
    // This is an arbitrary value to get regular blockhash updates for balance checks without  
    // hitting the RPC node with too many requests
    const BLOCKHASH_REFRESH_MILLIS: u64 = DEFAULT_MS_PER_SLOT * 32;        
    let mut latest_blockhash = client.get_latest_blockhash()?;               
    let mut now = Instant::now();                                 
    let mut fees = 0;
    for mut message in messages.iter().cloned() {
        if now.elapsed() > Duration::from_millis(BLOCKHASH_REFRESH_MILLIS) {
            latest_blockhash = client.get_latest_blockhash()?;
            now = Instant::now();          
        }
        message.recent_blockhash = latest_blockhash;                                           
        fees += client.get_fee_for_message(&message)?;   // We make RPC call per each message       
    }              
    Ok(fees)             
}       
```

For something similar to:

```rust
pub fn get_approximated_fees_for_messages(
    messages: &[Message],
    client: &RpcClient,
) -> Result<u64, Error> {
    let mut fees = 0;
    if messages.is_empty() {
        return Ok(fees);
    }
    // Get the latest blockhash as late as possible in the function.
    let latest_blockhash = client.get_latest_blockhash()?;
    // Get the first message fees as base
    let mut first_message = messages.get(0).unwrap().clone();
    first_message.recent_blockhash = latest_blockhash;
    let base_fee = client.get_fee_for_message(&first_message)?;
    
    messages.iter().cloned().for_each(|mut m| {
        m.recent_blockhash = latest_blockhash;
        fees += base_fee
    });
    Ok(fees)
}
```

Multiple observations here:

* If i understood correctly, the [transactions fees](https://docs.solana.com/implemented-proposals/transaction-fees) are deterministic. So getting the fee for the first one and multiplying could be a valid approach for this process, as all transactions should be the identical.
* Removed the previous regular blockhash acquisition, as now we are only doing 1 RPC call for getting the fees and setting such value for all the rest of the messages. So there should be no time to exceed  `400ms (slot time) x 32 = 12800ms`. I could be wrong here, though.
* If we only want this to affect to the `spl tokens distribution` feature, we could create a dedicated function `get_approximated_fees_for_messages` and use it [here](https://github.com/solana-labs/solana/blob/master/tokens/src/spl_token.rs#L99) , replacing the current one. But probably the normal token distribution feature should also get the benefit of this new way of calculating fees right ?

#### Use of idempotent accounts

Sorry my lack of context here. Are we talking about [read only accounts](https://docs.solana.com/developing/programming-model/accounts#read-only) ? How would that help in this case ?

Well, sorry the long comment. I would really appreciate yours ! Happy to move in the best direction for the project. Thanks in advance and have a wonderful week !

**Comment by norwnd** [comment]
Seems to be addressed and can be closed ? Just friendly reminder, @joncinque @brooksprumo

**Comment by joncinque** [comment]
Thanks for the ping! Resolved with #31419
